File: src/main/cpp/src/crema/core/SimpleDomain.cpp
#include "crema/core/SimpleDomain.h"
#include "crema/utils/ArraysUtil.h"

#include <algorithm>

using namespace crema::core;
using namespace std;

SimpleDomain::SimpleDomain(Domain *domain)
{
    this->variables = domain->getVariables();
    this->sizes = domain->getSizes();
    this->size = domain->getSize();
}

SimpleDomain::SimpleDomain(vector<int> *variables, vector<int> *sizes, int size)
{
    this->variables = variables;
    this->sizes = sizes;
    this->size = size;
}

int SimpleDomain::indexOf(int variable)
{
    // return crema::utils::indexOf(this->variables, variable); TODO:
    return 0;
}

int SimpleDomain::getCardinality(int variable)
{
    int offset = indexOf(variable);
    return sizes->at(offset);
}

bool SimpleDomain::contains(int variable)
{
    return indexOf(variable) >= 0;
}

vector<int> *SimpleDomain::getVariables()
{
    return variables;
}

vector<int> *SimpleDomain::getSizes()
{
    return sizes;
}

int SimpleDomain::getSize()
{
    return size;
}

int SimpleDomain::getSizeAt(int index)
{
    return sizes->at(index);
}

void SimpleDomain::removed(int variable)
{
    int index = -(indexOf(variable) + 1);
    for (; index < size; ++index)
    {
        variables->assign(index, variables->at(index) - 1);
    }
}

File: src/main/java/ch/idsia/crema/core/SimpleDomain.java
package ch.idsia.crema.core;

import java.util.Arrays;

/**
 * Domain implementation that does not include strides. You are likely to need {@link Strides} instead of this class.
 *
 * @author davidhuber
 */
public class SimpleDomain implements Domain {
	final private int[] variables;
	final private int[] sizes;
	final private int size;

	@Override
	public int getSize() {
		return size;
	}

	public SimpleDomain(Domain domain) {
		this.variables = domain.getVariables();
		this.sizes = domain.getSizes();
		size = variables.length;
	}

	public SimpleDomain(int[] variables, int[] sizes) {
		this.variables = variables;
		this.sizes = sizes;
		size = variables.length;
	}

	@Override
	public int indexOf(int variable) {
		return Arrays.indexOf(variables, variable);
	}

	@Override
	public boolean contains(int variable) {
		return indexOf(variable) >= 0;
	}

	@Override
	public int[] getVariables() {
		return variables;
	}

	@Override
	public int[] getSizes() {
		return sizes;
	}

	@Override
	public int getCardinality(int variable) {
		int offset = indexOf(variable);
		return sizes[offset];
	}

	@Override
	public int getSizeAt(int index) {
		return sizes[index];
	}

	@Override
	public void removed(int variable) {
		int index = -(indexOf(variable) + 1);
		for (; index < size; ++index) {
			--variables[index];
		}
	}

}
